docs/proposals/shoot-control-plane-migration/README.md

```diff
@@ -0,0 +1,234 @@
+
+It should not scale itself down, of course.
+
+##### State Diagram
+
+![State Diagram](state-diagram.png)
+
+This component should also expose a `/healthz` endpoint which returns healthy only if it can get the current lease status and if the current seed cluster is the owner of the lease for the given shoot. This `/healthz` endoint can be used for livenessProbe in the shoot machine-controller-manager, shoot etcd and also perhaps in the VPN components to avoid co-ordination problems in case the shoot control-plane co-ordinator itself crashes.
+
+###### Melt-down scenario
+
+In the [above](#shoot-control-plane-co-ordinator) approach, there is a possibility of a melt-down-like scenario where the co-ordinator is not able to access the lease status for some long period even though it actually owns the lease during that period, but the gardener is able to access the seed cluster. In such a scenario, co-ordinator will scale down the relevant component to zero while a gardener reconciliation will scale them back up to the original values. This cycle can potentially keep happening while the scenario persists.
+
+To address this issue, the co-ordinator can be supplied an additional source for the current lease status via some resource within the same namespace as the co-ordinator. The co-ordinator can choose to use the lease status from such a local resource or the remote/global/distributed resource depending on the validity of the lease periods in both locations. The gardener reconciliation can update the lease status on the local resource in the namespace of the co-ordinator.
+
+The remote/global/distributed source for lease status is still required to handle the scenario where the gardener is not able to access the shoot control-plane on the seed cluster.
+
+The assumption is that the opposite melt-down scenario of co-ordinator scaling up the control-plane components while gardener reconciliation scales it down will not happen because in such a scenario the gardener reconciliation would actually delete the shoot control-plane rather than just scale down.
+
+### Pros
+
+* No changes to existing shoot control-plane components except for additional livenessProbes.
+* Works for different mechanisms for advertising the current seed cluster lease for a shoot cluster.
+
+### Cons
+
+* One more additional component (shoot control-plane co-ordinator) in the shoot control-plane. Gardener reconciliation should be enhanced to handle the shoot control-plane co-ordinator.
+* Needs some sort of mechanism to advertise current seed cluster lease for a shoot cluster which need to be actively or passively kept up-to-date with changes to the shoot YAML in the garden cluster.
+
+### Changes Required in Gardener Reconciliation
+
+On movement of seed for a shoot control-plane,
+
+1. Update the lease status (this is optional if the lease status is calculated based on shoot YAML).
```

Comment:
Well, can't be the shoot resource, right? It's not reachable.

Reply:
Assuming we go with lease service (by no means, finalized), I think a persisted approach (rather than a calculated) where gardener reconciles it might be a safer bet.

The point of listing all the options was that each of them had some problems and there was no clear winning approach.

Reply:
Not sure what you mean. I would definitely not implement any lease service myself. Too much effort to get it right. Either we do DNS (preferred way - for me at least) or we let the Gardener write/update the lease in the seed cluster(s). The Gardener can't do that with normal reconciliation (happens not frequent enough), but a second controller could update the leases every 30s everywhere if necessary. However, this puts a lot of stress on the Gardener and strictly requires a super-HA Gardener, otherwise the shoot control planes will shut down within a minute if the Gardener is down. That is not desirable and since I have no hope to reach an availability on par with DNS ever (neither for the Gardener or any lease service we may implement ourselves), I would stick to DNS if we can make this work.

Reply:
Agreed that DNS would be preferable. We will evaluate.